### PR TITLE
[Java] Add Blob wrapper class around byte []

### PR DESCRIPTION
**Pull request by chadwalters**
Added a simple Blob wrapper class that allows for implementation of equals and
hashCode methods based on the contents of the blob.

Update codegen and blob example to reflect this change as well.

### CODE REVIEW

File: java/core/src/main/java/com/microsoft/bond/Blob.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * Provide access to underlying {@code byte []}
+     * @return the underying {@code byte []}
+     */
+    public byte[] getData()
+    {
+        return this.data;
+    }
+
+    /**
+     * Provide a shallow clone of this {@code Blob} instance. Underlying {@code byte []} will be shared
```
Comment (tstein) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: ~~ Do you need the second space in `{@code byte []}`?

File: java/core/src/main/java/com/microsoft/bond/Blob.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * Creates a new Blob instance wrapping an existing {@code byte []}
+     * @param data
+     * @return new Blob instance wrapping the data array
+     * @throws IllegalArgumentException if the argument is null
+     */
+    public Blob(byte[] data) {
+        if (data == null) {
+            throw new IllegalArgumentException("Argument 'data' must not be null");
+        }
```
Comment (sigansky) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: We have a helper to do that check: ArgumentHelper.ensureNotNull

File: java/core/src/main/java/com/microsoft/bond/Blob.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * Compare for equality by contents of underlying {@code byte []}
+     * @param other object to compare against
+     * @return {@code true} if object is also a Blob and the underlying arrays have equivalent contents
+     */
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
```
Comment (sigansky) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: Arrays.equals already takes care of this minor optimization.

File: java/core/src/main/java/com/microsoft/bond/Blob.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * Provide a shallow clone of this {@code Blob} instance. Underlying {@code byte []} will be shared
+     * @return a shallow clone of this {@code Blob} instance
+     */
+    @Override
+    protected Blob clone() {
+        return new Blob(this.data);
+    }
+
+    private byte[] data;
```
Comment (sigansky) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: >private [](start = 4, length = 7)

final

File: java/core/src/main/java/com/microsoft/bond/BlobBondType.java [hunk 5]
```diff
@@ -72,25 +72,25 @@ public final boolean isGenericType() {
...
-    protected final byte[] newDefaultValue() {
+    protected final Blob newDefaultValue() {
         return DEFAULT_VALUE;
     }
 
     @Override
-    protected final byte[] cloneValue(byte[] value) {
+    protected final Blob cloneValue(Blob value) {
         return value.clone();
     }
```
Comment (sigansky) [inline java/core/src/main/java/com/microsoft/bond/BlobBondType.java]: This should return a deep clone, which is not what Blob.clone does. Please see the API of cloneValue in BondType.

File: examples/java/core/blob/src/main/java/com/microsoft/bond/examples/BlobExample.java [hunk 6]
```diff
@@ -3,8 +3,10 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.ByteBuffer;
```
Comment (chwarr) [inline examples/java/core/blob/src/main/java/com/microsoft/bond/examples/BlobExample.java]: Minor: is this import needed anymore?

Comment (chadwalters) [inline examples/java/core/blob/src/main/java/com/microsoft/bond/examples/BlobExample.java]: Removed.

File: java/core/src/main/java/com/microsoft/bond/Blob.java (hunk 4 above)
Comment (chadwalters) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: Addressed.

File: java/core/src/main/java/com/microsoft/bond/BlobBondType.java (hunk 5 above)
Comment (chadwalters) [inline java/core/src/main/java/com/microsoft/bond/BlobBondType.java]: Addressed.

File: java/core/src/main/java/com/microsoft/bond/Blob.java (hunk 3 above)
Comment (chadwalters) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: Not exactly. The parameters to Array.equals() are the arrays -- this is the optimization at the level of Object.

File: java/core/src/main/java/com/microsoft/bond/Blob.java (hunk 2 above)
Comment (chadwalters) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: Addressed.

File: java/core/src/main/java/com/microsoft/bond/Blob.java (hunk 1 above)
Comment (chadwalters) [inline java/core/src/main/java/com/microsoft/bond/Blob.java]: Addressed throughout file.

**Comment by chadwalters** [comment]
CR comments address. Please squash when merging.

**Comment by chwarr** [comment]
Golden tests are failing. Can you re-gen?

**Pull request merged**